shared/src/PayloadCollection.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// The base class for all payloads that accept a <see cref="PayloadConvention"/>
+    /// </summary>
+    public abstract class PayloadCollection : IEnumerable<object>
+    {
+        /// <summary>
+        /// The underlying collection for the payload.
+        /// </summary>
+        public IDictionary<string, object> Collection { get; private set; } = new Dictionary<string, object>();
+
+        /// <summary>
+        /// The convention to use with this payload.
+        /// </summary>
+        public PayloadConvention Convention { get; internal set; }
+
+        /// <summary>
+        /// Get the value at the specified key
+        /// </summary>
+        /// <param name="key">Key of value</param>
+        /// <remarks>
+        /// This accessor is best used to access simple types. It is recommended to use <see cref="TryGetValue"/> to cast a complex type.
+        /// </remarks>
+        /// <returns>The specified property.</returns>
```

Comment:
...as an object.

What does a user do with the object after that? Will it cast? Do they need to deserialize? Only for complex types?

Reply:
we have this comment above:
"This accessor is best used to access simple types. It is recommended to use <see cref="TryGetValue"/> to cast a complex type."; would it help if we call out explicitly that they should cast simple types?

Reply:
Neither comment addresses simple types. I think it is worth mentioning.

Also, TryGetValue doesn't strictly cast, right? It will either cast or deserialize. Right?